pyro/distributions/transforms/generalized_channel_permute.py

```diff
@@ -0,0 +1,164 @@
+    lower triangular matrix with ones on the diagonal, and :math:`U\\sim C\\times C` is an upper triangular matrix.
+    :math:`W` is initialized to a random orthogonal matrix. Then, :math:`P` is fixed and the learnable parameters set
+    to :math:`L,U`.
+
+    The input :math:`\\mathbf{x}` and output :math:`\\mathbf{y}` both have shape `[...,C,H,W]`, where `C` is the number
+    of channels set at initialization.
+
+    This operation was introduced in [1] for Glow normalizing flow, and is also known as 1x1 invertible convolution.
+    It appears in other notable work such as [2,3], and corresponds to the class `tfp.bijectors.MatvecLU` of TensorFlow
+    Probability.
+
+    Example usage:
+
+    >>> channels = 3
+    >>> base_dist = dist.Normal(torch.zeros(channels, 32, 32), torch.ones(channels, 32, 32))
+    >>> inv_conv = GeneralizedChannelPermute(channels=channels)
+    >>> flow_dist = dist.TransformedDistribution(base_dist, [inv_conv])
+    >>> flow_dist.sample()  # doctest: +SKIP
+
+    :param channels: Number of channel dimensions in the input.
+    :type channels: int
+
+    1. Glow: Generative Flow with Invertible 1x1 Convolutions. [arXiv:1807.03039]
+    Diederik P. Kingma, Prafulla Dhariwal.
+
+    2. WaveGlow: A Flow-based Generative Network for Speech Synthesis. [arXiv:1811.00002]
+    Ryan Prenger, Rafael Valle, Bryan Catanzaro.
+
+    3. Neural Spline Flows. [arXiv:1906.04032]
+    Conor Durkan, Artur Bekasov, Iain Murray, George Papamakarios.
+
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 3
+
+    def __init__(self, channels=3):
+        super(GeneralizedChannelPermute, self).__init__(cache_size=1)
+        self.channels = channels
+
+        # Sample a random orthogonal matrix
+        W, _ = torch.qr(torch.randn(channels, channels))
+
+        # Construct the partially pivoted LU-form and the pivots
+        LU, pivots = W.lu()
+
+        # Convert the pivots into the permutation matrix
+        P, _, _ = torch.lu_unpack(LU, pivots)
+
+        # We register the permutation matrix so that the model can be serialized
+        self.register_buffer('permutation', P)
+
+        # NOTE: For this implementation I have chosen to store the parameters densely, rather than
+        # storing L, U, and s separately
+        self.LU = torch.nn.Parameter(LU)
+
+    def _call(self, x):
+        """
+        :param x: the input into the bijection
+        :type x: torch.Tensor
+
+        Invokes the bijection x=>y; in the prototypical context of a
+        :class:`~pyro.distributions.TransformedDistribution` `x` is a sample from the base distribution (or the output
+        of a previous transform)
+        """
+
+        # Extract the lower and upper matrices from the packed LU matrix
+        U = self.LU.triu()
+        L = self.LU.tril()
+        L.diagonal(dim1=-2, dim2=-1).fill_(1)
```

Comment:
You might want to avoid in-place ops here and elsewhere, e.g.
```py
L = self.LU.tril(-1) + torch.eye(self.LU.size(-1), dtype=self.LU.dtype, device=self.LU.device)
```
Actually I would be surprised if gradients worked with the in-place version. Do you have smoke tests ensuring gradients work?

Reply:
There is no error when I call `backward()` through a loss involving this transform and a transformed distribution, although I'll use your idiom to be safe